src/accelerate/deepspeed_utils.py

```diff
@@ -0,0 +1,177 @@
+    """
+
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+        self.optimizer = model.optimizer
+
+        self.progressive_layer_drop = model.progressive_layer_drop
+        self.global_steps = model.global_steps
+
+        # overwriting micro_steps for user's gradient_accumulation
+        self.model.micro_steps = -1
+
+    def forward(self, *args, **kwargs):
+        return self.model(*args, **kwargs)
+
+    def train(self, *args, **kwargs):
+        return self.model.train(*args, **kwargs)
+
+    def eval(self, *args, **kwargs):
+        return self.model.eval(*args, **kwargs)
+
+    def is_gradient_accumulation_boundary(self):
+        return self.model.is_gradient_accumulation_boundary()
+
+    def zero_optimization(self):
+        return self.model.zero_optimization()
+
+    def amp_enabled(self):
+        return self.model.amp_enabled()
+
+    def fp16_enabled(self):
+        return self.model.fp16_enabled()
+
+    def _take_model_step(self, *args, **kwargs):
+        return self.model._take_model_step(*args, **kwargs)
+
+    def step(self, lr_kwargs=None):
+        """ DeepSpeedEngine.step() without `micro_steps` update & no profiling """
+        if self.is_gradient_accumulation_boundary(): # it shouldn't matter whether we keep this line or not
+            if self.progressive_layer_drop:
+                self.progressive_layer_drop.update_state(self.global_steps)
+
+            self._take_model_step(lr_kwargs)
+
+    def backward(self, loss):
+        """ DeepSpeedEngine.backward() with no loss scaling & no profiling """
+        if self.zero_optimization():
+            self.optimizer.is_gradient_accumulation_boundary = self.is_gradient_accumulation_boundary(
+            )
+            self.optimizer.backward(loss)
+        elif self.amp_enabled():
+            assert is_apex_available, "You have enabled apex in deepspeed_plugin, but apex is unavailable in your machine"
+            # AMP requires delaying unscale when inside gradient accumulation boundaries
+            # https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations
+            delay_unscale = not self.is_gradient_accumulation_boundary()
+            with amp.scale_loss(loss,
+                                self.optimizer,
+                                delay_unscale=delay_unscale) as scaled_loss:
+                scaled_loss.backward()
+        elif self.fp16_enabled():
+            self.optimizer.backward(loss)
+        else:
+            loss.backward()
+
+        # this will ensure deepspeed gradient_accumulation matches user's accumulation
+        self.model.micro_steps += 1
+
+    def save_checkpoint(self, *args, **kwargs):
+        return self.model.save_checkpoint(*args, **kwargs)
+
+    def load_checkpoint(self, *args, **kwargs):
+        return self.model.load_checkpoint(*args, **kwargs)
+
+
+class DeepSpeedOptimizerWrapper(torch.optim.Optimizer):
```

Comment:
Should this subclass the `AcceleratedOptimizer` for most properties and only overrides the necessary ones?